src/main/java/spoon/support/util/ModelList.java

```diff
@@ -0,0 +1,192 @@
+		this.owner = owner;
+	}
+
+	protected abstract CtRole getRole();
+	protected abstract int getDefaultCapacity();
+	protected abstract void onSizeChanged(int newSize);
+
+	@Override
+	public T get(int index) {
+		return list.get(index);
+	}
+
+	@Override
+	public int size() {
+		return list.size();
+	}
+
+	@Override
+	public T set(int index, T element) {
+		T oldElement = list.get(index);
+		if (oldElement == element) {
+			//no change
+			return oldElement;
+		}
+		ensureModifiableStatementsList();
+		getModelChangeListener().onListDelete(owner, getRole(), list, index, oldElement);
+		element.setParent(owner);
+		getModelChangeListener().onListAdd(owner, getRole(), list, index, element);
+		list.set(index, element);
+		return oldElement;
+	}
+
+	@Override
+	public boolean contains(Object o) {
+		return list.contains(o);
+	}
+	@Override
+	public boolean isEmpty() {
+		return list.isEmpty();
+	}
+
+	@Override
+	public Object[] toArray() {
+		return list.toArray();
+	}
+
+	@Override
+	public <T> T[] toArray(T[] a) {
+		return list.toArray(a);
+	}
+
+	@Override
+	public boolean add(T e) {
+		if (e == null) {
+			return false;
+		}
+		ensureModifiableStatementsList();
+		e.setParent(owner);
+		getModelChangeListener().onListAdd(owner, getRole(), list, e);
+		boolean result = list.add(e);
+		onSizeChanged(list.size());
+		return result;
+	}
+
+	@Override
+	public boolean remove(Object o) {
+		if (list.isEmpty()) {
+			return false;
+		}
+		int size = list.size();
+		for (int i = 0; i < size; i++) {
+			//first do not use equals, but same
+			if (list.get(i) == o) {
+				remove(i);
+				return true;
+			}
+		}
+		int idx = list.indexOf(o);
+		if (idx >= 0) {
+			remove(idx);
+			return true;
+		}
+		return false;
+	}
+
+	@Override
+	public boolean containsAll(Collection<?> c) {
+		return list.containsAll(c);
+	}
+
+	@Override
+	public void clear() {
+		getModelChangeListener().onListDeleteAll(owner, getRole(), list, new ArrayList<>(list));
+		list = CtElementImpl.emptyList();
+		onSizeChanged(list.size());
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		return list.equals(o);
+	}
+
+	@Override
+	public int hashCode() {
+		return list.hashCode();
+	}
+
+	@Override
+	public void add(int index, T element) {
+		if (element == null) {
+			return;
+		}
+		ensureModifiableStatementsList();
+		element.setParent(owner);
+		getModelChangeListener().onListAdd(owner, getRole(), list, index, element);
+		list.add(index, element);
+		onSizeChanged(list.size());
+	}
+
+	@Override
+	public T remove(int index) {
+		T oldElement = list.get(index);
+		getModelChangeListener().onListDelete(owner, getRole(), list, index, oldElement);
+		list.remove(index);
+		onSizeChanged(list.size());
+		return oldElement;
+	}
+
+	@Override
+	public int indexOf(Object o) {
+		return list.indexOf(o);
+	}
+
+	@Override
+	public int lastIndexOf(Object o) {
+		return list.lastIndexOf(o);
+	}
+
+	private void ensureModifiableStatementsList() {
```

Comment:
This method should be changed to be generic

Reply:
why? It is private and it uses generic argument <T> from ListModel<T> class

Reply:
Actually no you the name says it's for statements and you have this: `list == CtElementImpl.<CtStatement>emptyList()`

It should use a `T` and have a name like `ensureModifiableList`

Reply:
Now I see it ;-), Thanks. I will fix it probably today in the evening